File: javascript/5.最长回文子串.js
/*
 * @lc app=leetcode.cn id=5 lang=javascript
 *
 * [5] 最长回文子串
 */

// @lc code=start
/**
 * @param {string} s
 * @return {string}
 */
var longestPalindrome = function(s) {
    if (s.length == 1) return s

    var res = s[0];
    for (var idx=0; idx<s.length; idx++) {
        var i=idx-1, j=idx+1;

        // 如果中间都是一样的字符，则这些肯定能组成回文串
        while (i>-1 && s[i]==s[idx]) {
            i--;
        }
        while (j<s.length && s[j]==s[idx]) {
            j++;
        }

        while (i>-1 && j<s.length && s[i]==s[j]) {
            i--;
            j++;
        }
        
        if (j-i-1 > res.length) {
            res = s.slice(i+1, j);
        }
    }

    return res;
};
// @lc code=end

File: java/5.最长回文子串.java
/*
 * @lc app=leetcode.cn id=5 lang=java
 *
 * [5] 最长回文子串
 */

// @lc code=start
class Solution {
    public String longestPalindrome(String s) {
        String res = Character.toString(s.charAt(0));
        for (int i=0; i<s.length(); ++i) {
            int left=i, right=i;
            while (right+1<s.length() && s.charAt(right+1) == s.charAt(i)) { 
                ++right; 
            }
            while (left-1>-1 && right+1<s.length() && s.charAt(left-1)==s.charAt(right+1)) {
                --left;
                ++right;
            }
            if (right-left+1 > res.length()) {
                res = s.substring(left, right+1);
            }
        }
        return res;
    }
}
// @lc code=end

